src/services/GLSL/vertex_shader_catalog.glsl

```diff
@@ -0,0 +1,212 @@
+#define SQRT 2
+#define SQUARE 3
+#define POWER 4
+#define GAMMA 5
+
+uniform sampler2D uCmapTexture;
+uniform sampler2D uXTexture;
+uniform sampler2D uYTexture;
+uniform sampler2D uSizeTexture;
+uniform sampler2D uColorTexture;
+uniform sampler2D uOrientationTexture;
+uniform highp usampler2D uSelectedSourceTexture;
+uniform sampler2D uSizeMinorTexture;
+
+uniform vec2 uFrameViewMin;
+uniform vec2 uFrameViewMax;
+uniform vec3 uPointColor;
+uniform highp int uShapeType;
+uniform int uNumCmaps;
+uniform int uCmapIndex;
+uniform float uFeatherWidth;
+uniform float uPointSize;
+uniform bool uSizeMajorMapEnabled;
+uniform bool uAreaMode;
+uniform bool uShowSelectedSource;
+uniform bool uSizeMinorMapEnabled;
+uniform bool uAreaModeMinor;
+uniform bool uCmapEnabled;
+uniform bool uOmapEnabled;
+
+uniform float uRotationAngle;
+uniform vec2 uRangeOffset;
+uniform vec2 uRangeScale;
+uniform float uScaleAdjustment;
+
+out vec3 v_colour;
+out float v_pointSize;
+out float v_orientation;
+out float v_selected;
+out float v_minorSize;
+out float v_featherWidth;
+
+
+vec4 getValueByIndexFromTexture(sampler2D texture, int index) {
+    ivec2 size = textureSize(texture, 0);
+    int row = index / size.x;
+    int col = index - row * size.x;
+    return texelFetch(texture, ivec2(col, row), 0);
+}
+
+uvec4 getValueByIndexFromTexture(usampler2D texture, int index) {
+    ivec2 size = textureSize(texture, 0);
+    int row = index / size.x;
+    int col = index - row * size.x;
+    return texelFetch(texture, ivec2(col, row), 0);
+}
+
+vec2 imageToGL(vec2 imageVec) {
+    return 2.0 * imageVec - 1.0;
+}
+
+vec3 hsv2rgb(vec3 c)
+{
+    vec4 K = vec4(1.0, 2.0 / 3.0, 1.0 / 3.0, 3.0);
+    vec3 p = abs(fract(c.xxx + K.xyz) * 6.0 - K.www);
+    return c.z * mix(K.xxx, clamp(p - K.xxx, 0.0, 1.0), c.y);
+}
+
+float getSquareSideByArea(float area, float minorSize) {
+    switch (uShapeType) {
+        case BOX_FILLED:
+        case BOX_LINED:
+        case RHOMB_FILLED:
+        case RHOMB_LINED:
+        case CROSS_FILLED:
+        case CROSS_LINED:
+        case X_FILLED:
+        case X_LINED:
+        case LineSegment_FILLED:
+            return sqrt(area);
+        case CIRCLE_FILLED:
+        case CIRCLE_LINED:
+            return sqrt(area / PI) * 2.0;
+        case HEXAGON_FILLED:
+        case HEXAGON_LINED:
+        case HEXAGON_FILLED_2:
+        case HEXAGON_LINED_2:
+            return sqrt((2.0 * area) / (3.0 * SQRT3)) * SIN_60 * 2.0;
+        case TRIANGLE_FILLED_UP:
+        case TRIANGLE_LINED_UP:
+        case TRIANGLE_FILLED_DOWN:
+        case TRIANGLE_LINED_DOWN:
+            return sqrt(area * 4.0 / SQRT3);
+        case ELLIPSE_FILLED:
+        case ELLIPSE_LINED:
+            float side = sqrt(area / PI) * 2.0;
+            if (minorSize >= 0.0 && area < minorSize) {
+                side = sqrt(minorSize / PI) * 2.0;
+            }   
+            return side;
+        default:
+            return 0.0;
+    }
+}
+
+bool isNaN(float val) {
```

Comment:
checked /services/GLSL/pixel_shader_raster.glsl and the definition is different (return val <= -FLT_MAX; in pixel_shader_raster.glsl), will this cause problem?

Reply:
good point. The other definition is due to a bug in the Apple shader compiler on iOS, if I remember correctly. I don't think we need to worry about it for now, but we should test on iPad to see if this issue crops up with catalogs as well